speaker/layer2_controller.go

```diff
@@ -41,20 +42,43 @@ func (c *layer2Controller) SetConfig(log.Logger, *config.Config) error {
 // The speakers parameter is a map with the node name as key and the readiness
 // status as value (true means ready, false means not ready).
 // If the speakers map is nil, it is ignored.
-func usableNodes(eps *v1.Endpoints, speakers map[string]bool) []string {
+func usableNodes(eps k8s.EpsOrSlices, speakers map[string]bool) []string {
 	usable := map[string]bool{}
-	for _, subset := range eps.Subsets {
-		for _, ep := range subset.Addresses {
-			if ep.NodeName == nil {
-				continue
-			}
-			if speakers != nil {
-				if ready, ok := speakers[*ep.NodeName]; !ok || !ready {
+	switch eps.Type {
+	case k8s.Eps:
+		for _, subset := range eps.EpVal.Subsets {
+			for _, ep := range subset.Addresses {
+				if ep.NodeName == nil {
 					continue
 				}
+				if speakers != nil {
+					if ready, ok := speakers[*ep.NodeName]; !ok || !ready {
+						continue
+					}
+				}
+				if _, ok := usable[*ep.NodeName]; !ok {
+					usable[*ep.NodeName] = true
+				}
 			}
-			if _, ok := usable[*ep.NodeName]; !ok {
-				usable[*ep.NodeName] = true
+		}
+	case k8s.Slices:
+		for _, slice := range eps.SlicesVal {
+			for _, ep := range slice.Endpoints {
+				if ep.Conditions.Ready == nil || !*ep.Conditions.Ready {
+					continue
+				}
+				nodeName := ep.Topology["kubernetes.io/hostname"]
```

Comment:
ditto, topology is going to be removed soon https://github.com/kubernetes/kubernetes/commit/9613f79adde3e0080fe25576cb37a74f25262848 , use `ep.NodeName`